teps/0057-tekton-catalog-pipeline-organization.md

```diff
@@ -0,0 +1,470 @@
+
+```yaml=
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: foo-pipeline
+  labels:
+    app.kubernetes.io/version: '0.2'
+  annotations:
+    tekton.dev/pipelines.minVersion: '0.12.1'
+    tekton.dev/displayName: foo-pipeline
+spec:
+  tasks:
+    - name: fetch-repo
+      taskRef:
+        name: git-clone
+        version: '0.1'
+        catalog:
+          - name: tektoncd
+        url: http://raw.githubusercontent.com/tektoncd/catalog/task/foo/1.0/foo.yaml
+```
+
+Pros:
+
+1. This would help the user to easily understand from which catalog the task is used
+2. There won't be any requirement to add the tasks in the `/tasks` directory
+
+Cons:
+
+1. This would require an API change
+
+- (Solution 3) --- By adding a config which describes the resources in the pipeline
+
+This adds a an additional config which has basic information(for e.g. name, version, url, etc)
+of the resources used in the pipeline
+
+Folder structure will be like:
+
+```bash=
+
+./pipeline/                     👈 The kind of the resource
+
+    /p1                         👈 Definition file must have same name
+       /OWNERS                  👈 owners of this resource
+       /0.1
+         /README.md
+         /install-script.sh     👈 To install all resources related to pipeline
+         /p1.yaml               👈 The file name should match the resource name
+         /samples/
+         /tasks                 👈 Tasks which pipelines refers
+         /reference
+           /tasks-reference.yaml   👈 manifest design which gives basic info of resources used in the pipeline
+         /support
+       /0.2/...
+
+    /p2
+       /OWNERS
+       /README.md
+       /0.1
+         /README.md
+         /p2.yaml
+         /samples/
+         /tasks
+```
+
+The structure of the manifest will be like:
+
+```yaml
+catalogs:
+  - name: tektoncd
+    url: github.com/tektoncd/catalog
+  - name: openshift
+    url: github.com/openshift/pipelines-catalog
+
+tasks:
+  - name: gke-deploy
+    version: '0.1'
+    catalog:
+      name: tektoncd
+
+  - name: kaniko
+    version: '0.1'
+    catalog:
+      name: openshift
+```
+
+- This can help the tools like `tkn-hub` to install the tasks used in the pipeline
+
+For example:
+
+```bash=
+$ tkn hub install pipeline build-and-push-gke-deploy
+
+This will install the following:
+    * gke-deploy task - tekton catalog
+    * kaniko task - openshift catalog
+
+Do you really want to install the pipeline and related resources mentioned above? (y/N):
+```
+
+Pros:
+
+1. This would help the users to understand how pipeline is structured and
+   which tasks are used in the catalog and from whtere
+2. This would help the tools like `tkn-hub` to install the tasks used
+   in the pipeline
+
+Cons:
+
+- This would help the user to understand easily which tasks are used in the pipeline
+- Since it is just an informative one which would just describe how pipeline
```

Comment:
Since the manifest just includes the information about the tasks and from which catalog they belong, hence it won't install the dependent tasks right neither the pipeline right ??? To install the pipeline, the user will have to apply the pipeline manifest which will install the pipeline only right ??